grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -716,58 +900,66 @@ def _GetArraySchema(items_type_name: str) -> ArraySchema:
   }
 
 
-def _GetFieldSchema(
+def _GetMapEntryTypeName(field_name: str) -> str:
+  """Extract the name of the associated map type from a field's name."""
+  capitalized_name_components = [
+    comp.capitalize() for comp in field_name.split("_")
+  ]
+
+  return f"{''.join(capitalized_name_components)}Entry"
+
+
+def _GetMapFieldKeyValueTypes(
     field_descriptor: FieldDescriptor,
-) -> Union[SchemaReference, ArraySchema, OneofFieldSchema]:
-  """Create a dictionary to be used as the schema of a message field.
+) -> Tuple[Optional[FieldDescriptor], Optional[FieldDescriptor]]:
```

Comment:
The type of your function is `Tuple[Optional[X], Optional[Y]]` but in reality, you only return the key if and only if you return  the value. You could make this more explicit by having the following type: `Optional[Tuple[X, Y]]`.

Reply:
That is correct, but using `Optional[Tuple[X, Y]]` would require that the returned value of this function to be stored in a variable to check whether the value `is None` before accessing the tuple elements as required by static type checking. This would imply that either:
- we don't use sequence unpacking anymore and rather index the new variable used to store the returned tuple (or `None`);
- we continue to use sequence unpacking, but use a temporary variable only to check whether the function returned `None` or not.

Which of the 3 approaches (3rd being the one used right now) do you consider to be the best?

Reply:
How about defining a `KeyValueDescriptor` which is a named tuple with `key` and `value` properties? Then you can use the first approach, where you don't unpack, but you don't have to use ugly and confusing indexing.

Reply:
Awesome idea! Thank you for your suggestion, I implemented it.